File: Solutions by Programming Task/Fibonacci sequence/Go/TailRecursive.go
package main

import "fmt"

func FibImpl(p int, n int, i int) int {
	if i == 0 {
		return n
	}
	return FibImpl(p + n, p, i - 1)
}

func Fib(x int) int {
	return FibImpl(1, 0, x)
}

func main() {
	fmt.Println(Fib(10))
}

File: Solutions by Programming Task/Fibonacci sequence/Rust/TailRecursive.rs
fn fib_impl(p: usize, n: usize, i: usize) -> usize {
    match i {
        0 => p,
        _ => fib_impl(n, p + n, i - 1)
    }
}

fn fib(x: usize) -> usize {
    fib_impl(0, 1, x)
}

fn main() {
    println!("{}", fib(10))
}
